### PR TITLE
fix(es/compat): Preserve array rest when lowering nested object rest

### LINKED ISSUE #12339: panic on object rest nested inside an array rest binding
### Describe the bug

SWC crashes when an object-rest binding exists anywhere within the target pattern of an array rest element.
All of the following patterns cause SWC to panic when targeting ES5:
```js
const [...{ ...y }] = [1, 2];
const [...{ 0: { ...y } }] = [{ a: 1 }]; 
const [...[{ ...y }]] = [{ a: 1 }];
```

### Input code

```typescript
const [...{ ...y }] = [1, 2];
console.log(y);
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false
    },
    "target": "es5",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Link to the code that reproduces this issue

https://play.swc.rs/?version=1.16.2&code=H4sIAAAAAAAAA0vOzysuUYjW09OrVgASlQq1sQq2CtGGOgpGsda8XMlA6fycVL2c%2FHSNSk1rAFszL0suAAAA&config=H4sIAAAAAAAAA1WPSw7DIAwF9zkF8rrbdtE79BCIOhERP9mOVBTl7oUE0maH3xszsA5KwcwGnmotxzIkTYx0ziXhHER%2FSgKSE7IhmwRuvRWu1agd4x5tRwOiaUKpW8j3hoOLkbHjLfM22DH%2FC030iZD5ClZUh8nhVTc0Jfj4XvayfaQ%2B9tA%2F4Ad12XkxWH71TaEFh%2B0LYuVI0xQBAAA%3D

### SWC Info output

_No response_

### Expected behavior

Babel 8.0.1 lowers the same input to:

```js
function _objectDestructuringEmpty(t) {
  if (null == t) throw new TypeError("Cannot destructure " + t);
}
function _extends() {
  return (
    (_extends = Object.assign
      ? Object.assign.bind()
      : function (n) {
          for (var e = 1; e < arguments.length; e++) {
            var t = arguments[e];
            for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]);
          }
          return n;
        }),
    _extends.apply(null, arguments)
  );
}
var y = _extends({}, (_objectDestructuringEmpty([1, 2]), [1, 2]));
console.log(y); // {0: 1, 1: 2}
```

### Actual behavior

panicked at crates/swc_ecma_compat_es2015/src/destructuring.rs:156:30:
internal error: entered unreachable code: rest pattern should handled by array pattern handler: Rest(RestPat { span: 2382..2395, dot3_token: 2382..2385, arg: Array(ArrayPat { span: 2385..2395, elems: [Some(Object(ObjectPat { span: 2386..2394, props: [Rest(RestPat { span: 2388..2392, dot3_token: 2388..2391, arg: Ident(BindingIdent { id: Ident { span: 2391..2392, ctxt: #2, sym: "y", optional: false }, type_ann: None }), type_ann: None })], optional: false, type_ann: None }))], optional: false, type_ann: None }), type_ann: None })

### Version

1.16.2

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**

Fix a panic when lowering patterns such as `const [...{ ...y }] = [1, 2]` to ES5.

Preserve the array rest element when splitting the pattern, replacing only its argument with a temporary binding. This ensures remaining elements are collected before lowering the nested object rest and prevents emitting an invalid standalone rest declaration.

Preserve the original RHS value throughout destructuring assignments and as their expression result, even when the pattern or its side effects overwrite the source binding.

This PR is scoped to array-rest structure and assignment-result preservation in the ES2018 object-rest transform. Existing limitations in loop-head TDZ preservation and parameter `arguments` mapping are outside its scope.

Add fixture and execution coverage for direct and nested patterns, assignments, function parameters, for-of bindings, and iterator consumption.

**Related issue (if exists):**

- Closes #12339.

### CODE REVIEW

File: crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs [hunk 1]
```diff
@@ -857,6 +857,11 @@ impl<O: RestOutput> RestLowerer<O> {
 
         // Extract nested pattern and replace with temp
         let nested_pat: Pat = match arr.elems[split_idx].as_mut() {
+            Some(Pat::Rest(rest)) => {
+                let nested = rest.arg.take();
+                *rest.arg = split_ref.clone().into();
+                *nested
```
[the remark this answers is not in the document]
Comment (magic-akari) [reply crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:841]: Valid issue. This PR fixes nested object-rest lowering within array rest and preserves assignment results. Loop-head TDZ loss predates this change and also affects ordinary object-rest patterns. Babel 8.0.1 exhibits the same issue, both with object-rest lowering alone and with an ES5 target.

[the remark this answers is not in the document]
Comment (magic-akari) [reply crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:841]: Valid issue. This PR fixes nested object-rest lowering within array rest and preserves assignment results. The mapped `arguments` issue predates this change and also affects ordinary object-rest parameters. Babel 8.0.1's ES5 transform likewise returns `2` instead of `1` here. I suggest addressing parameter semantics separately, including subsequent ES2015 lowering.

File: crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs [hunk 2]
```diff
@@ -165,33 +165,11 @@ impl VisitMutHook<TraverseCtx> for ObjectRestSpreadPass {
...
-                AssignExpr {
-                    span: DUMMY_SP,
-                    left: ref_ident.clone().into(),
-                    op: op!("="),
-                    right: right.take(),
-                }
-                .into(),
-            );
-        }
+        let ref_ident = out.declare_temp("_ref");
```
[the remark this answers is not in the document]
Comment (magic-akari) [reply crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:171]: Valid issue. The underlying temporary-scoping limitation already exists on main for call RHS expressions, but this PR introduces a regression for identifier RHS expressions by capturing them unconditionally. The example below passes on main and fails with this PR.

This PR addresses nested object-rest lowering within array rest and assignment-result preservation. We acknowledge the regression and are deferring the parameter-initializer temporary-scoping fix.

For comparison, Babel and esbuild also fail this example when lowering object rest while preserving default parameters:

```js
let value, rest;
const outer = { a: undefined, b: "outer" };
const inner = { a: 1, b: "inner" };

function reenter() {
    f(inner);
    return 0;
}

function f(
    source,
    result = ({ a: value = reenter(), ...rest } = source)
) {
    return [result === source, rest.b];
}

console.log(JSON.stringify(f(outer)));
```

Native execution and SWC main’s object-rest transform print `[true,"outer"]`.

- [Babel 8.0.5 playground](https://babeljs.io/repl#?config_lz=N4IgZglgNgpgdgQwLYxALhAJxgBygOgCsBnEAGhDwFcBzCOUtAbSZAAEAjBDmKAemp04AWgAumBAzAB7TEmHSOhGAGNRw7MXXEc2BABNywAL4BdUxWLSqmFTAAqATxyoMxFZgg5RIY0A&code_lz=DYUwLgBAbghsCuIA0EBOIDOYDcAoAxgPYB2WEh8YIqEAvBAN4QwBcE8xAJiAGYCWxEJxQAjNgCIKVVOIgBfPEVKQBgmvSasIARlETV1WQty4eHfGD4k0IEMWkAKAJSNcEdxB4ODqJ3g82YPCoxBAADHhyJmbEFlahXm4eGBSo-MhJ7ugY8MCQ9A6abLAIIHQ2do5OKAB0ddmQcuUpwelOuC4MmYHBoQDa2bn5tPQtacg2WDUiALqRJkopoDXAhADmDgBSAMoA8gByNVioAmt8PACeDl5S1E732EA&lineWrap=true&version=8.0.5), using only `@babel/plugin-transform-object-rest-spread`: `[false,"inner"]`.
- [esbuild 0.28.2 playground](https://esbuild.github.io/try/#dAAwLjI4LjIALS10YXJnZXQ9ZXMyMDE3AGxldCB2YWx1ZSwgcmVzdDsKY29uc3Qgb3V0ZXIgPSB7IGE6IHVuZGVmaW5lZCwgYjogIm91dGVyIiB9Owpjb25zdCBpbm5lciA9IHsgYTogMSwgYjogImlubmVyIiB9OwoKZnVuY3Rpb24gcmVlbnRlcigpIHsKICAgIGYoaW5uZXIpOwogICAgcmV0dXJuIDA7Cn0KCmZ1bmN0aW9uIGYoCiAgICBzb3VyY2UsCiAgICByZXN1bHQgPSAoeyBhOiB2YWx1ZSA9IHJlZW50ZXIoKSwgLi4ucmVzdCB9ID0gc291cmNlKQopIHsKICAgIHJldHVybiBbcmVzdWx0ID09PSBzb3VyY2UsIHJlc3QuYl07Cn0KCmNvbnNvbGUubG9nKEpTT04uc3RyaW5naWZ5KGYob3V0ZXIpKSk7), targeting ES2017: `[true,"inner"]`. The assignment result is preserved, but rest extraction still uses the inner call’s source.
- This PR’s SWC object-rest transform: `[false,"inner"]`.

Comment (kdy1) [inline crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:171]: [P2] Keep parameter-initializer captures per invocation

Parameter defaults are visited before the function body opens a VarDeclarations scope, so this queued var can be emitted in an enclosing scope. If an accessor/default re-enters the function while destructuring, the inner call overwrites _ref; the outer rest extraction and assignment result then use the inner RHS. Capture this evaluation in a per-invocation scope (or otherwise avoid an enclosing shared binding).

Comment (kdy1) [inline crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:171]: [P2] Avoid writing the generated temp through a with environment

In a non-strict with block, _ref resolves against the with object before the injected var binding. A scope object with an _ref setter can therefore observe or reject this new assignment for ({ a, ...rest } = source), whereas the identifier-RHS fast path previously performed no _ref write. Use a capture strategy that cannot be intercepted by the active with environment.

Comment (kdy1) [inline crates/swc_ecma_transformer/src/es2018/object_rest_spread.rs:171]: [P2] Do not expose the capture to direct eval

A default such as a: value = eval("_ref = replacement") now resolves to this emitted binding. That changes _ref before the later lowerer steps, so the generated rest value and assignment result come from replacement rather than the original source (and strict-mode source behavior can change from a ReferenceError to success). The old identifier fast path had no such binding.

**Review by kdy1 — COMMENTED** [review]
Review Full completed against base 90f880906c648a9877797d250f7f692895818ee2 and head 0deab796fd074eeabca405cfeff57fcb4e2c9065.

Three baseline reviews completed successfully (four worker attempts, including one retry after incomplete output). Six follow-up passes retired reviewer 1 in pass 3, reviewer 2 in pass 5, and reviewer 3 in pass 6.

Findings:
- [P2] The unconditional RHS capture is shared with parameter-initializer evaluation and can be overwritten by reentrant execution.
- [P2] The generated capture is observable through non-strict with environment resolution.
- [P2] A direct eval in a destructuring default can mutate the generated capture before rest extraction completes.

The review covered the pinned diff, transformer hook/declaration scopes, directive handling, nested rest lowering, fixtures, and execution-order edge cases. Material limitation: the reviewers did not run Cargo tests; one baseline attempt produced incomplete CLI output and was retried successfully.

**Comment by kdy1** [comment]
If you don't think these points worth fixing, you can skip it (and feel free to do it)

**Review by kdy1 — APPROVED** [review]
